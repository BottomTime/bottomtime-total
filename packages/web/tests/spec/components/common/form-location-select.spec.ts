import {
  ComponentMountingOptions,
  flushPromises,
  mount,
} from '@vue/test-utils';

import FormLocationSelect from '../../../../src/components/common/form-location-select.vue';

const TestId = 'mah-select';
const Latitude = 25.915738052691395;
const Longitude = -80.07950492939037;

const RadiusSlider = `[data-testid="${TestId}-radius"]`;
const ChangeLocationButton = `[data-testid="${TestId}-select-btn"]`;
const ClearLocationButton = `[data-testid="${TestId}-clear-btn"]`;
const CoordinatesText = `[data-testid="${TestId}-coords"]`;

// Location dialog
const LatitudeInput = 'input#latitude';
const LongitudeInput = 'input#longitude';
const ConfirmLocationButton = '[data-testid="confirm-location"]';
const CancelLocationButton = '[data-testid="cancel-location"]';

describe('FormLocationSelect component', () => {
  let opts: ComponentMountingOptions<typeof FormLocationSelect>;

  beforeAll(() => {
    opts = {
      global: {
        stubs: { teleport: true },
      },
    };
  });

  it('will render correctly with no value', () => {
    const wrapper = mount(FormLocationSelect, opts);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('will render correctly without radius slider', () => {
    const wrapper = mount(FormLocationSelect, {
      props: {
        modelValue: {
          lat: Latitude,
          lon: Longitude,
        },
      },
      ...opts,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('will render correctly with radius slider', () => {
    const wrapper = mount(FormLocationSelect, {
      props: {
        modelValue: {
          lat: Latitude,
          lon: Longitude,
          radius: 80,
        },
        showRadius: true,
      },
      ...opts,
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('will allow a user to select a new location', async () => {
    const wrapper = mount(FormLocationSelect, {
      props: {
        showRadius: false,
        testId: TestId,
      },
      ...opts,
    });

    await wrapper.get(ChangeLocationButton).trigger('click');
    await wrapper.get(LatitudeInput).setValue(Latitude);
    await wrapper.get(LongitudeInput).setValue(Longitude);
    await wrapper.get(ConfirmLocationButton).trigger('click');
    await flushPromises();

    expect(wrapper.find(ConfirmLocationButton).exists()).toBe(false);
    expect(wrapper.get(CoordinatesText).text()).toBe(
      `${Latitude}, ${Longitude}`,
    );

    expect(wrapper.emitted('update:modelValue')).toEqual([
      [
        {
          lat: Latitude,
          lon: Longitude,
        },
      ],
    ]);
  });

  it('will allow a user to select a new location with a radius', async () => {
    const radius = 120;
    const wrapper = mount(FormLocationSelect, {
      props: {
        showRadius: true,
        testId: TestId,
      },
      ...opts,
    });

    await wrapper.get(ChangeLocationButton).trigger('click');
    await wrapper.get(LatitudeInput).setValue(Latitude);
    await wrapper.get(LongitudeInput).setValue(Longitude);
    await wrapper.get(ConfirmLocationButton).trigger('click');
    await flushPromises();

    await wrapper.get(RadiusSlider).setValue(radius);

    expect(wrapper.find(ConfirmLocationButton).exists()).toBe(false);
    expect(wrapper.get(CoordinatesText).text()).toBe(
      `${Latitude}, ${Longitude}`,
    );

    const emitted = wrapper.emitted('update:modelValue');
    expect(emitted).toHaveLength(2);
    expect(emitted![1]).toEqual([
      {
        lat: Latitude,
        lon: Longitude,
        radius,
      },
    ]);
  });

  it('will allow a user to update a location', async () => {
    const radius = 120;
    const wrapper = mount(FormLocationSelect, {
      props: {
        showRadius: true,
        testId: TestId,
        modelValue: {
          lat: 77.7,
          lon: -7.77,
          radius: 200,
        },
      },
      ...opts,
    });

    await wrapper.get(ChangeLocationButton).trigger('click');
    await wrapper.get(LatitudeInput).setValue(Latitude);
    await wrapper.get(LongitudeInput).setValue(Longitude);
    await wrapper.get(ConfirmLocationButton).trigger('click');
    await flushPromises();

    await wrapper.get(RadiusSlider).setValue(radius);

    expect(wrapper.find(ConfirmLocationButton).exists()).toBe(false);
    expect(wrapper.get(CoordinatesText).text()).toBe(
      `${Latitude}, ${Longitude}`,
    );

    const emitted = wrapper.emitted('update:modelValue');
    expect(emitted).toHaveLength(2);
    expect(emitted![1]).toEqual([
      {
        lat: Latitude,
        lon: Longitude,
        radius,
      },
    ]);
  });

  it('will allow a user to cancel changing a location', async () => {
    const wrapper = mount(FormLocationSelect, {
      props: {
        modelValue: {
          lat: Latitude,
          lon: Longitude,
        },
        testId: TestId,
      },
      ...opts,
    });

    await wrapper.get(ChangeLocationButton).trigger('click');
    await wrapper.get(CancelLocationButton).trigger('click');

    expect(wrapper.find(ConfirmLocationButton).exists()).toBe(false);
    expect(wrapper.get(CoordinatesText).text()).toBe(
      `${Latitude}, ${Longitude}`,
    );
  });

  it('will allow a user to clear a location', async () => {
    const wrapper = mount(FormLocationSelect, {
      props: {
        showRadius: true,
        testId: TestId,
        modelValue: {
          lat: 77.7,
          lon: -7.77,
          radius: 200,
        },
      },
      ...opts,
    });

    await wrapper.get(ClearLocationButton).trigger('click');

    expect(wrapper.find(CoordinatesText).exists()).toBe(false);
    expect(wrapper.find(RadiusSlider).exists()).toBe(false);
    expect(wrapper.emitted('update:modelValue')).toEqual([[undefined]]);
  });
});
