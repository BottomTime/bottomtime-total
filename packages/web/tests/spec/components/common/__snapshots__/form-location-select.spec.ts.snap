// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLocationSelect component will render correctly with no value 1`] = `
"<teleport-stub to="#modal">
  <transition-stub name="backdrop" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
  <!--v-if-->
</teleport-stub>
<!--v-if-->
<div class="flex gap-2"><button class="text-grey-950 p-2 m-0 text-sm border-1 border-grey-800 shadow-sm shadow-grey-800 rounded-md disabled:text-grey-500 bg-secondary hover:bg-secondary-hover" type="button">
    <!--v-if--><span class="text-xs">Select location... </span>
  </button>
  <!--v-if-->
</div>"
`;

exports[`FormLocationSelect component will render correctly with radius slider 1`] = `
"<teleport-stub to="#modal">
  <transition-stub name="backdrop" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
  <!--v-if-->
</teleport-stub>
<div class="text-sm my-2">
  <p class="space-x-2"><span class="text-danger"><i class="fa-solid fa-map-marker-alt fa-fw"></i></span><label class="font-bold">Coordinates</label></p>
  <p class="ml-6"><span>25.915738052691395, -80.07950492939037</span></p>
  <p class="space-x-2"><span><i class="fa-solid fa-ruler-horizontal fa-fw"></i></span><label class="font-bold">Range:</label></p>
  <p class="ml-6 flex gap-2 items-center">
  <div class="flex gap-2 items-baseline align-baseline w-full"><input class="grow appearance-none h-2 rounded-full bg-grey-300 dark:bg-grey-400 cursor-pointer accent-blue-500 hover:dark:accent-blue-400 hover:accent-blue-600" type="range" min="10" max="500" step="10">
    <!--v-if-->
  </div><span class="text-nowrap text-right min-w-10">80 km </span></p>
</div>
<div class="flex gap-2"><button class="text-grey-950 p-2 m-0 text-sm border-1 border-grey-800 shadow-sm shadow-grey-800 rounded-md disabled:text-grey-500 bg-secondary hover:bg-secondary-hover" type="button">
    <!--v-if--><span class="text-xs">Change location... </span>
  </button><button class="text-grey-950 p-2 m-0 text-sm border-1 border-grey-800 shadow-sm shadow-grey-800 rounded-md disabled:text-grey-500 bg-secondary hover:bg-secondary-hover" type="button">
    <!--v-if--><span class="text-xs">Clear location</span>
  </button></div>"
`;

exports[`FormLocationSelect component will render correctly without radius slider 1`] = `
"<teleport-stub to="#modal">
  <transition-stub name="backdrop" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
  <!--v-if-->
</teleport-stub>
<div class="text-sm my-2">
  <p class="space-x-2"><span class="text-danger"><i class="fa-solid fa-map-marker-alt fa-fw"></i></span><label class="font-bold">Coordinates</label></p>
  <p class="ml-6"><span>25.915738052691395, -80.07950492939037</span></p>
  <!--v-if-->
</div>
<div class="flex gap-2"><button class="text-grey-950 p-2 m-0 text-sm border-1 border-grey-800 shadow-sm shadow-grey-800 rounded-md disabled:text-grey-500 bg-secondary hover:bg-secondary-hover" type="button">
    <!--v-if--><span class="text-xs">Change location... </span>
  </button><button class="text-grey-950 p-2 m-0 text-sm border-1 border-grey-800 shadow-sm shadow-grey-800 rounded-md disabled:text-grey-500 bg-secondary hover:bg-secondary-hover" type="button">
    <!--v-if--><span class="text-xs">Clear location</span>
  </button></div>"
`;
